### DISCUSSION TITLE
sqlalchemy.exc.InvalidRequestError: Can't operate on closed transaction inside context manager.  Please complete the context manager before emitting further commands.

### DISCUSSION THREAD

**Discussion by mark-barrett** [sqlalchemy/sqlalchemy#9114] [Usage Questions] [answered]
We have a function that originally checked for an existence of a unique record and if it existed, update, else create.

We found that if we received multiple requests at around the same time (specifically around webhooks), this would break as between the moment one worker checks to see if the record exists and it creating it, another would have already created it.

To get around this, I put the create first in a try catch, caught the IntegrityError, rolled back and then updated instead. Something similar to this:

```python
try:
    object = Model(**some_dict)
    db.add(object)
    await db.commit()
    await db.refresh(object)
except IntegrityError:
    await db.rollback()
    result = await db.execute(select(Model).where(Model.some_id == some_value))
    object = result.scalar_one()
    object.some_field_to_update = some_other_value
    await db.commit()
``` 

the entire thing is passed the `db` object from a dependency in the FastAPI path, something like

```python
async with SessionLocal() as session:
    yield session
    await session.commit()
```

This all works fine. When I debug the code, it does exactly what I expected it to do, however the problem comes when I run the already existing unit tests.

In our unittests we use a Pytest fixture which looks like this:

```python
async with db_engine.begin() as connection:
    await connection.run_sync(Base.metadata.drop_all)
    await connection.run_sync(Base.metadata.create_all)
    async with SessionLocal(bind=connection) as session:
        yield session
        await session.flush()
        await session.rollback()
```

But when I run the tests I get the error:

`sqlalchemy.exc.InvalidRequestError: Can't operate on closed transaction inside context manager.  Please complete the context manager before emitting further commands.`

I wouldn't normally post a question if I can find the answer online, but I Googled for this and I can't find anything on it. I can deduce roughly what it means, but I'm a little bit confused as to why this issue is only presenting itself now. Nothing much has changed in the actual business logic.

I'm wondering perhaps is it a direct affect of something I did in the code? We have been using that FastAPI depends db object on that endpoint for months and have no issues.

Is it something to do with some session being closed and it not using the right one?

Any help would be greatly appreciated.

**Comment by zzzeek** [comment, marked as answer]
The exception occurs if you run a "with engine.begin()" block and then commit or rollback the transaction inside the block, which ordinarily you should not be doing (maybe a rollback only, the commit is automatic), and then you try to emit more commands inside the block (or I can see that perhaps simply calling .commit() in the block might trigger this itself when using ORM sessions, not sure).

I think it's likely because you are calling "db.rollback" in your custom context manager, then running more commands.  That rollback will roll back the outer transaction that you started with db.begin(), which means you can't emit more SQL commands on the connection within the block.

if you have a test suite where you want tests to be within an enclosing transaction that's rolled back, but then you also want to call "rollback" within tests themselves, you have to use savepoints, where the Session will take ownership of a savepoint as the thing it rolls back, and not your enclosing transaction.

We have a much better system for this in 2.0, however in 1.4, the pattern for this is documented at https://docs.sqlalchemy.org/en/14/orm/session_transaction.html#joining-a-session-into-an-external-transaction-such-as-for-test-suites .

**Reply by mark-barrett** [reply]
So the business logic/API seems to work fine. The only thing I needed to worry about is to expunge any objects I might need before I do the rollback. Continuing after the rollback to query seems to work fine when I test it manually. Does that sounds correct?

Is it just the test suite that would be affected? If so, why?

I checked out the docs link. Am I correct in saying that by setting it up the way it's stated there, each time we make a transaction, once the transaction is ended, we are setting a savepoint that we can roll back to? Am I correct in saying the part of the example that really makes it come together is the event part with the calling of begin_nested? Everything else seems fairly standard.

Also, does this work with Async? Tried a few examples but no luck getting it to work.

**Reply by mark-barrett** [reply]
Also, might be unrelated, by while on the topic. In the example I gave, regardless of if there was a result created or found in the except block, after that, we add a particular other object to a relationship on that instance.

e.g. after the try, catch above, the model instance might have something like this done to it:

```python
object.people.append(person)
await db.commit()
```

This gives a `sqlalchemy.exc.InvalidRequestError: Can't attach instance <Person at 0x7f7c3662ed30>; another instance with key (<class 'app.models.fingerprint.Person'>, (1,), None) is already present in this session.`, not only for this particular append, but any append that occurs thereafter the rollback.

Is there something else I need to add to the session?

**Reply by zzzeek** [reply]
> So the business logic/API seems to work fine. The only thing I needed to worry about is to expunge any objects I might need before I do the rollback. Continuing after the rollback to query seems to work fine when I test it manually. Does that sounds correct?

from this end "test it manually" is a little ambiguous, if you are in an engine.begin() block, then rollback(), then run more SQL commands, it will fail.  If "test manually" is not using that block, then you wouldnt get the error.


> Is it just the test suite that would be affected? If so, why?

I'm just able to tell you what the error means, as far as test suite vs. not that's something I am looking at.

> 
> I checked out the docs link. Am I correct in saying that by setting it up the way it's stated there, each time we make a transaction, once the transaction is ended, we are setting a savepoint that we can roll back to? 

yes

> Am I correct in saying the part of the example that really makes it come together is the event part with the calling of begin_nested? Everything else seems fairly standard.

the event part is what makes it start a new savepoint if you rollback the old savepoint.   SQLAlchemy 2.0 has a new feature where you no longer need to use an event handler.

> Also, does this work with Async? Tried a few examples but no luck getting it to work.

it should work in theory but I haven't tested it.   you need to associate events with the Session that's proxied by AsyncSession, there are some examples at https://docs.sqlalchemy.org/en/14/orm/extensions/asyncio.html#using-events-with-the-asyncio-extension  .  there might be other issues if using sqlite (I wouldn't recommend asyncio with sqlite).

**Reply by zzzeek** [reply]
> e.g. after the try, catch above, the model instance might have something like this done to it:
> 
> ```python
> object.people.append(person)
> await db.commit()
> ```
> 
> This gives a `sqlalchemy.exc.InvalidRequestError: Can't attach instance <Person at 0x7f7c3662ed30>; another instance with key (<class 'app.models.fingerprint.Person'>, (1,), None) is already present in this session.`, not only for this particular append, but any append that occurs thereafter the rollback.

that "person" object is detached and coming from somewhere else and you already have person(id=1) in that session.  you can use `new_person = session.merge(person)` instead of `session.add(person)` to reconcile  that.

**Reply by mark-barrett** [reply]
Thank you so much for all your help, I really appreciate it. We will look to move towards v2 in the coming weeks/months.

In case anybody is looking for it, I managed to modify the original Pytest Fixture to use Savepoints:

```python
@pytest.fixture()
async def db_session(db_engine) -> AsyncSession:
    async with db_engine.begin() as connection:
        await connection.begin_nested()
        await connection.run_sync(Base.metadata.drop_all)
        await connection.run_sync(Base.metadata.create_all)

        async with SessionLocal(bind=connection) as session:
            @event.listens_for(
                session.sync_session, "after_transaction_end"
            )
            def end_savepoint(session, transaction):
                if connection.closed:
                    return

                if not connection.in_nested_transaction():
                    connection.sync_connection.begin_nested()

            yield session
            await session.flush()
            await session.rollback()
```

**Reply by zzzeek** [reply]
is that working?  would be good to know the event part of it works w/ the async proxies

**Reply by mark-barrett** [reply]
Yea works fine, had a few prints in the event listener too to check and it indeed gets called multiple times throughout the tests execution.

With the merge recommendation you made, the tests now pass! Thanks again!

Can give more information if needed.
